app/ide-desktop/lib/dashboard/src/authentication/src/authentication/providers/session.tsx

```diff
@@ -26,18 +27,52 @@ const SessionContext = react.createContext<SessionContextType>(
 // =======================
 
 interface SessionProviderProps {
+  /** URL that the content of the app is served at, by Electron.
+   *
+   * This **must** be the actual page that the content is served at, otherwise the OAuth flow will
+   * not work and will redirect the user to a blank page. If this is the correct URL, no redirect
+   * will occur (which is the desired behaviour).
+   *
+   * The URL includes a scheme, hostname, and port (e.g., `http://localhost:8080`). The port is not
+   * known ahead of time, since the content may be served on any free port. Thus, the URL is
+   * obtained by reading the window location at the time that authentication is instantiated. This
+   * is guaranteed to be the correct location, since authentication is instantiated when the content
+   * is initially served. */
+  mainPageUrl: URL;
+  registerAuthEventListener: listen.ListenFunction;
   userSession: () => Promise<results.Option<cognito.UserSession>>;
   children: react.ReactNode;
 }
 
-// eslint-disable-next-line @typescript-eslint/naming-convention
 export function SessionProvider(props: SessionProviderProps) {
-  const { children, userSession } = props;
+  const { mainPageUrl, children, userSession, registerAuthEventListener } =
+    props;
 
   /** Flag used to avoid rendering child components until we've fetched the user's session at least
    * once. Avoids flash of the login screen when the user is already logged in. */
   const [initialized, setInitialized] = react.useState(false);
 
+  /** Produces a new object every time.
+   * This is not equal to any other empty object because objects are compared by reference.
+   * Because it is not equal to the old value, React re-renders the component. */
+  function newRefresh() {
+    return {};
+  }
+
+  /** State that, when set, forces a refresh of the user session. This is useful when a
+   * user has just logged in (so their cached credentials are out of date). Should be used via the
+   * `refreshSession` function. */
+  const [refresh, setRefresh] = react.useState(newRefresh());
+
+  /** Forces a refresh of the user session.
+   *
+   * Should be called after any operation that **will** (not **might**) change the user's session.
+   * For example, this should be called after signing out. Calling this will result in a re-render
+   * of the whole page, which is why it should only be done when necessary. */
+  const refreshSession = () => {
```

Comment:
consider wrapping all the refresh stuff in an IIFE that returns just `[refresh, refreshSession]` - using `newRefresh` or `setRefresh` are both mistakes